t/helper/test-gvfs-protocol.c

```diff
@@ -0,0 +1,1191 @@
+	 *     "GET /<uri-base>/gvfs/prefetch?lastPackTimestamp=123456789 HTTP/1.1"
+	 *
+	 */
+	gvfs = strstr(uri_target, "/gvfs/");
+	if (!gvfs) {
+		logerror("could not parse GVFS REST API in '%s'", uri_target);
+		return -1;
+	}
+	strbuf_add(&req_data->uri_base, uri_target, (gvfs - uri_target));
+	strbuf_trim_trailing_dir_sep(&req_data->uri_base);
+
+	gvfs += 6; /* skip "/gvfs/" */
+	strbuf_add(&req_data->gvfs_api, "gvfs/", 5);
+	while (*gvfs && *gvfs != '/' && *gvfs != '?')
+		strbuf_addch(&req_data->gvfs_api, *gvfs++);
+
+	/*
+	 * We do attempt to split the query-params within the args.
+	 * The caller can do that if they need to.
+	 */
+	if (*gvfs == '/')
+		strbuf_addstr(&req_data->slash_args, gvfs + 1);
+	else if (*gvfs == '?')
+		strbuf_addstr(&req_data->quest_args, gvfs + 1);
+
+	/*
+	 * Read the set of HTTP headers into a string-list.
+	 */
+	while (1) {
+		if (strbuf_getwholeline_fd(&h, fd, '\n') == EOF)
+			goto done;
+		strbuf_trim_trailing_newline(&h);
+
+		if (!h.len)
+			goto done; /* a blank line ends the header */
+
+		string_list_append(&req_data->header_list,
+				   strbuf_detach(&h, NULL));
+	}
+
+	/*
+	 * We do not attempt to read the <message-body>, if it exists.
+	 * We let our caller read/chunk it in as appropriate.
+	 */
+done:
+	if (trace2_is_enabled())
+	{
+		struct string_list_item *item;
+		trace2_printf("%s: %s", TR2_CAT, req_data->start_line.buf);
+		for_each_string_list_item(item, &req_data->start_line_fields)
+			trace2_printf("%s: Field: %s", TR2_CAT, item->string);
+		trace2_printf("%s: [uri-base '%s'][gvfs '%s'][args '%s' '%s']",
+			      TR2_CAT,
+			      req_data->uri_base.buf,
+			      req_data->gvfs_api.buf,
+			      req_data->slash_args.buf,
+			      req_data->quest_args.buf);
+		for_each_string_list_item(item, &req_data->header_list)
+			trace2_printf("%s: Hdrs: %s", TR2_CAT, item->string);
+	}
+
+	return 0;
+}
+
+static int dispatch(struct my_request_data *req_data)
+{
+	const char *method = req_data->start_line_fields.items[0].string;
+
+	if (!strcmp(req_data->gvfs_api.buf, "gvfs/objects")) {
+
+		if (!strcmp(method, "GET"))
+			return do__gvfs_objects__get(req_data);
+		if (!strcmp(method, "POST"))
+			return do__gvfs_objects__post(req_data);
+	}
+
+	logerror("TODO unsuppored GVFS Protocol REST API ('%s','%s')",
+		 method, req_data->gvfs_api.buf);
+	return -1;
+}
+
+static int execute(void)
+{
+	struct my_request_data req_data = MY_REQUEST_DATA__INIT;
+	char *addr = getenv("REMOTE_ADDR"), *port = getenv("REMOTE_PORT");
+	int err = 0;
+
+	if (addr)
+		loginfo("Connection from %s:%s", addr, port);
+
+	set_keep_alive(0);
+
+	while (1) {
+		my_request_data__release(&req_data);
+
+		alarm(init_timeout ? init_timeout : timeout);
+		err = my_request_data__read(&req_data, 0);
+		alarm(0);
+
+		if (err == 0)
+			err = dispatch(&req_data);
+
+		if (err == -1)
+			break;
+
+		if (err == 1) { /* let EOF exit without error */
+			err = 0;
+			break;
+		}
```

Comment:
If you reversed the roles of `0` and `1`, i.e. if `0` denoted `EOF`, you could simply write `if (err < 1) break;` here.

Reply:
i've overhauled this.